README.md

```diff
@@ -1,19 +1,144 @@
+
+Take as an example the process of `extracting only odd numbers from the elements of a slice and returning those numbers by 3 times`.
+
+```golang
+func Odd3(s []int) []int {
+    var r []int
+    for _, v := range s {
+        if v%2 == 0 {
+            continue
+        }
+        r := append(v*3)
+    }
+    return r
+}
+```
+
+When using gcf, implement as follows.
+
+```golang
+// var s []int
+
+itb := gcf.FromSlice(s)
+itb = gcf.Filter(itb, func(v int) bool {
+    return v%2 > 0
+})
+itb = gcf.Map(itb, func(v int) int {
+    return v * 3
+})
+
+// Get the processing result as a slice
+r := itb.ToSlice(itb)
+```
+
+This example is meant to show how to use it briefly.  
+Replacing inline processing with gcf will significantly reduce the performance, so it is not recommended to rewrite processing that can be easily implemented and managed inline using gcf.
+
+## Environment
+
+- Go 1.18 RC (or Beta)
+
+Since gcf uses Generics, this version is currently in Beta status, but you need version 1.18 to use it.  
+We also have a container usage environment for vscode that you can use if you do not want to install Go 1.18 in your local environment.  
+(See below [.devcontainer] (https://github.com/meian/gcf/tree/main/.devcontainer))
+
+## Installation
+
+Install by using `go get` on the directory under the control of the Go module.
 
 ```bash
 go get -d github.com/meian/gcf
 ```
 
-# Performance
+## Design
+
+### Implements by Iterator
+
+gcf is designed to composite processing with the `Iterator` pattern.
+Some processes may allocate memory internally, but most processes avoid unnecessary memory allocations in the middle of the process.
+
+### Iterable + Iterator
+
+Each function returns `Iterable[T]`, which only has the ability to generate `Iterator[T]` by `Iterator()`.
+`Iterator[T]` moves the element position that collection returns by `MoveNext()` to next, and gets current element by `Current()`.
+Functions is composited by `Iterable[T]`, and the state is keep only in the `Iterator[T]`, which makes it easy to reuse the generated composition.
+
+### MoveNext + Current
+
+In Iterator implementation, you may see set of functions that uses `HasNext()` to check for the next element and `Next()` to move to the next element and return the element.
+In gcf, we implemented that `MoveNext()` moves to the next element and returns move result, and `Current()` returns the current element.
+This is because we took advantage to get current value multiple times without changing, rather than providing next check without changing.
+
+### Top-level functions
+
+In libraries of collection operations in other languages, the processing is often defined by methods so that we can use method chain, but we implemented by top-level functions.
+This is because generics in Go cannot define type parameters at the method level, so some functions cannot be provided by methods, and if only some functions are provided by methods, the processing cannot be maintained consistently.
```

Comment:
```suggestion
This is because generics in Go cannot define type parameters at the method level, so some functions cannot be provided by methods, and if only some functions are provided by methods, the processing cannot be maintained consistency.
```